spec/javascripts/modules/inline-attachment-modal-spec.js

```diff
@@ -0,0 +1,161 @@
+
+  function itBehavesLikeALinkAction (action) {
+    var url = 'https://example.com/' + action
+    var body = '<a href="' + url + '" data-modal-action="' + action + '">Link</a>'
+
+    it('updates the modal with the link response', function (done) {
+      inlineAttachmentModal.$modal.open()
+      inlineAttachmentModal.workflow.renderSuccess({ body: body })
+
+      fetchMock.get(url, '<h1>Link followed</h1>')
+      var dynamicSection = inlineAttachmentModal.$modal.querySelector('.js-dynamic-section')
+      dynamicSection.querySelector('a').click()
+
+      fetchMock.flush(true).then(function () {
+        expect(dynamicSection).toBeVisible()
+        expect(dynamicSection).toContainHtml('<h1>Link followed</h1>')
+        done()
+      })
+    })
+  }
+
+  beforeEach(function () {
+    buildModalDialogue()
+
+    link = document.createElement('a')
+    link.dataset.modalAction = 'open'
+    link.href = 'https://example.com/inline-attachment'
+
+    inlineAttachmentModal = new InlineAttachmentModal(link)
+    inlineAttachmentModal.init()
+  })
+
+  afterEach(function () {
+    removeModalDialogue()
+  })
+
+  describe('open action', function () {
+    it('opens when the link is clicked and shows the content', function (done) {
+      fetchMock.get('https://example.com/inline-attachment', '<h1>Inline attachment</h1>')
+      link.click()
+
+      fetchMock.flush(true).then(function () {
+        var dynamicSection = inlineAttachmentModal.$modal.querySelector('.js-dynamic-section')
+        expect(dynamicSection).toBeVisible()
+        expect(dynamicSection).toContainHtml('<h1>Inline attachment</h1>')
+        done()
+      })
+    })
+  })
+
+  // describe('insert action', function () {
+  //   it('inserts the markdown into the editor and closes the modal', function () {
+  //     var markdown = '[Image:file.jog]'
+  //     var body = '<button type="button" data-modal-action="insert" data-modal-data="' + markdown + '">Insert</button>'
+  //     inlineAttachmentModal.$modal.open()
+  //     inlineAttachmentModal.workflow.renderSuccess({ body: body })
+  //     spyOn(inlineAttachmentModal.editor, 'insertBlock')
+  //
+  //     inlineAttachmentModal.$modal.querySelector('button').click()
+  //     expect(inlineAttachmentModal.$modal).toBeHidden()
+  //     expect(inlineAttachmentModal.editor.insertBlock).toHaveBeenCalledWith(markdown)
+  //   })
+  // })
+
+  describe('upload action', function () {
+    itBehavesLikeAFormAction('upload')
+  })
+
+  describe('insert action', function () {
+    itBehavesLikeALinkAction('insert')
+  })
+
+  describe('delete action', function () {
+    itBehavesLikeAFormAction('delete')
+  })
+
+  describe('edit action', function () {
+    itBehavesLikeALinkAction('edit')
+  })
+
+  describe('update action', function () {
+    itBehavesLikeAFormAction('update')
+  })
+
+  describe('back action', function () {
+    itBehavesLikeALinkAction('back')
+  })
+
+  describe('insert-attachment-block action', function () {
```

Comment:
Never mind: the attachment one is (necessarily) even more inconsistent!

Reply:
Sorry I didn't quite understand this. Were you saying never mind to your own comment? 

This test is more complex as we've wrote our own logic in the file being tested, whereas the simple ones are the ones that utilise modalWorkflow.

Reply:
Indeed I was. Sorry to confuse. I understand the reason for the different; the idea was to make the helper functions a little more flexible, so they could accommodate all the modal actions. But it's just an idea I thought worth mentioning in case you thought it might be a significant improvement.

Reply:
No problem and thanks for the suggestion. My stance is that if we have to write basically this same code again (the application code rather than the test) for the next modal we introduce then we should actually have it handled in modalWorkflow rather than in the individual modalClass which itself should prompt making the tests more generic or to test they delegate.

I see them as on rules of 3 where we have 2 examples now and when the third one comes it'll be time to act.